ls1intum/Artemis
# Refactor services on the server

This PR starts with a refactoring of server services and their dependencies
Fixes #2867 

### Checklist

### Motivation and Context
We have too many (circular) dependencies between resources and services. We should instead follow the design principles `low coupling and high cohesion`

### Description
* Reduce potential circular dependencies (in particular between exam, exercises and course)
* Move simple service methods (e.g. `findByIdElseThrow`) into repositories as `default` methods (Example: before this PR almost every service and resource class had dependencies to `UserService`, now almost all only depend on the simple `UserRepository` for authentication checks)
* Extract smaller services with a dedicated purpose (e.g. `ExamDateService`)
* Move methods into different services in which they make more sense (e.g. in the exam services)
* I also moved some files into different folders (to make the code structure easier to understand, in case we have too many merge conflicts with other PRs, I could also revert this)

### Steps for Testing
Basically **all areas** are affected, so we should click through the most important use cases.
I tried to be careful so that no functionality is changed

### Test Coverage
no changes (might be slightly higher as I deleted some unused code)

### Screenshots
no changes
